### ISSUE TITLE
Removing current Fl_Group does not reset Fl_Group::current_

### ISSUE THREAD

**Issue by turix2** [fltk/fltk#88]
If an `Fl_Group` object is being removed from the view hierarchy and happens to be the "current" group (for example, the most recently added group), then `Fl_Group::current_` is left pointing to an invalid object.

This will cause problems, for example, in the `Fl_Widget` constructor, which has `if (Fl_Group::current()) Fl_Group::current()->add(this);`.  The next time a widget is created after removing the current group will cause a crash.

There seem to be a few options:

- Leave as is, but document things better.  (In particular, show examples of "safe removal patterns" for group widgets.)
- Add something like `if (Fl_Group::current_ == this) Fl_Group::current(NULL)` or `if (Fl_Group::current_ == this) this->end();` to the `Fl_Group::~Fl_Group()` destructor.
    - I would prefer the former here because reassigning the pointer to parent risks recreating the same problem, but maybe this can be avoided by being careful with removal orders in destructors?
- Make `Fl_Group::current_` a `std::weak_ptr` (or something similar).  Of course, I realize this would have huge ripple effects.  I'm just putting it our there in case you were already contemplating an upgrade to smart pointers anyway.

**Comment by Albrecht-S** [comment]
Thanks for the report. To clarify: how do you remove the `Fl_Group` widget "from the view hierarchy"? For testing and analyzation this would be an important information.

Can you post a small, self-contained (i.e. complileable with fltk-config) code example that shows the issue? I believe that the case you describe is an unusual case, but I don't say it's not a valid case to care about.

Note that you can remove a widget "from the view hierarchy" just by calling `parent()->remove()` w/o actually destroying the widget, but this would at least leave `Fl_Group::current()` pointing at a valid widget. The important case seems indeed to be when an `Fl_Group` widget gets `delete`'d and I'd also prefer the first case you proposed, to add `if (Fl_Group::current_ == this) Fl_Group::current(NULL)` to the destructor. Not a final solution though, just my gut feeling right now.

And yes, no smart pointers in FLTK, we stick to way older standards (as you may know, i.e. no std:: and not C++11 or later).

**Comment by erco77** [comment]
Regarding smart pointers, we do have that Fl_Widget_Tracker thing I think Matthias added. Could that help here?

**Comment by Albrecht-S** [comment]
(History: The `Fl_Widget_Tracker` class was added by me, but its internal mechanism had been created by Matthias before. It was a team work.)

Thanks for the heads-up, but I can't see how this could help here. `Fl_Widget_Tracker` is designed to "watch" a widget during a short code sequence, particularly when calling a callback. The `Fl_Widget_Tracker` instance should be deleted when it's no longer needed because it uses a linked list to track the widgets. Every widget that is deleted has to search the list to see if there's a "watcher", which is very ineffective for long lists [1]. So, as for watching a static widget from its creation until its deletion in a "static" d'tor, well that's really out of scope.

[1] sure, we could improve this, using a hashtable or a tree, but that's another story.

**Comment by turix2** [comment]
In this case, the Fl_Group object got removed using `delete`.   

I was removing several widgets at once, including parents and children.  But even though I know that FLTK would work if I just deleted everything once as a whole from the root of the hierarchy, my code works its way up the hierarchy deleting children first before their parents.  Then other (groups of) widgets are added.

I know my use case is a little "non-standard" because I'm dynamically adding and removing custom widgets (classes of my own, derived from Fl_Group, hence my desire to use `delete`), and I don't make use of the `current()` pattern at all.   

Since the Fl_Group constructor automatically calls `begin()` (something I should have known, but discovered while tracking down my crash, and the "real source" of my issue), I worked around my problem by adding a `Fl_Group::current(NULL);` call to my derived-class constructor.  (I realize that calling `begin()` in the constructor is convenient for most of the normal use cases, I'm just weird.)  

Anyway, even though I found a workaround, I figured it would still be good to let you know about the risk of the dangling pointer.

My code pretty big/hairy/complicated so I can't post it all here.  I'm happy to describe things more though if the above isn't clear enough.

And yeah, after looking a bit more at the FLTK source code and alternatives, I agree that just assigning `Fl_Group::current_` to NULL in the destructor seems like the easiest and safest solution.  Thanks!

**Comment by Albrecht-S** [comment]
Thanks for your comments. You're not alone: when I was using FLTK as a (commercial) user I did things much like you describe. My application had a client/server architecture where the server sent descriptions of widget trees and the client built whole windows with widgets, sent input results back, and got a new set of widgets to populate one or more windows. Of course the old widgets had to be removed. [Someone else is now maintaining this application, it's still alive AFAICT.]

At that time (1.1.5+ ... 1.3.0) there were issues with FLTK accessing deleted widgets after a callback and this was fixed by thorough usage of `Fl_Widget_Tracker` (mentioned above) and [the advice to use] `Fl::delete_widget()` to delete widgets not immediately when called but "after the event loop returns". [This has some issues with embedded event loops, obviously, but this is OT for now.]

Why am I writing all this? My hope is that `Fl::delete_widget()` is no longer necessary and you can delete widgets whenever you like (i.e. inside a callback, even in the widget's own callback). I'd appreciate if you could tell me (us) more about your experiences:

1. are you really (always) using `operator delete` directly to destroy widgets inside callbacks, i.e. as a result of a button click?

2. Is the button a user clicks or another widget that causes an event/callback also removed (deleted) in some or all cases?

3. Do you use Fl::delete_widget() at all? (I hope your answer is  'no')

4. Are there or have there been other issues you had to work around, related to your way to handle and delete widgets?

This is all interesting for further FLTK development and documentation, for instance to answer the question: "do we still recommend `Fl::delete_widget()` or is everything working fine with `delete` now?".

Thanks in advance, looking forward to your comments.

**Comment by turix2** [comment]
Haha!  Your first paragraph is very close to what we are doing right now.  We generate abstractions of "view" objects in a core layer that are serialized and sent to FLTK, from which interaction messages are sent back to our core. 

I'll answer your specific questions, but we also just mocked up a simple demo of the bug to attach that will hopefully give you the most clarity.  [fltk_example.cpp.txt](https://github.com/fltk/fltk/files/4775313/fltk_example.cpp.txt)

> are you really (always) using operator delete directly to destroy widgets inside callbacks, i.e. as a result of a button click?

We do always use operator delete, and it is triggered by user events like button clicks, but this does not happen inside of callbacks.  Rather, the callback sends a message is sent to our "core", which then may reply with one or more "removeView" messages (among other things).

> Is the button a user clicks or another widget that causes an event/callback also removed (deleted) in some or all cases?

Yes, sometimes.

> Do you use Fl::delete_widget() at all? (I hope your answer is 'no')

No, never.

> Are there or have there been other issues you had to work around, related to your way to handle and delete widgets?

I don't think there's anything else related to deleting.  We have had some other issues but most of that is probably more from getting used to FLTK than any issues with the code.  

Really appreciate the fast & detailed response here!  We'll keep track of further things in the future and let you know as we run into anything.  
Thanks again!!

**Comment by Albrecht-S** [comment]
Thanks for your detailed reply and the test code. I'll look into it tomorrow, but a first look at it shows that your case seems to be different because your threaded application working with `Fl::awake()` triggers the FLTK core outside the event loop, i.e. the `Fl::awake()` handling is not directly related to the user event. The user event (button click and its callback) terminates, the event loop is run, and eventually your thread calls `Fl::awake()`.

Awesome code, BTW, you did a great job to create such a demo program.

One thing I noticed in your demo in the "workaround" code (may be in your original code or not):
```
        /* For our workaround, we always disable the begin()/end()
         * functionality and explicitly reset Fl_Group::current */
        this->end();
        Fl_Group::current(NULL);
```
In these lines `this->end();` is redundant/useless, you can remove it. If you look at [the docs](https://www.fltk.org/doc-1.4/classFl__Group.html#a029653337efa0cb2ec52852519293ba8) you'll see that this "translates" to `Fl_Group::current(this->parent());` which is overwritten in the next statement. Nothing really bad, but an unnecessary instruction (maybe optimized away by the compiler anyway).

**Comment by Albrecht-S** [comment]
PS: is `#include <FL/../test/threads.h>` the `threads.h` in the FLTK distribution? I guess yes, but just to be sure. Too late here to test right now...

**Comment by turix2** [comment]
Thanks!  
Re your question about the `#include`, the answer is yes. 
However, I don't think the threads stuff shouldn't matter.  For example, the following (much simpler) code also demonstrates the problem:
```
// Simpler demonstration of FLTK Case 88 issue
// Set this to 1 to enable the workaround we are currently using.
#define FLTK_CASE_88_WORKAROUND 0
#include <assert.h>
#include <FL/Fl.H>
#include <FL/Fl_Window.H>
int
main(void)
{
    Fl_Window window(100, 100);
    Fl_Group * v2 = new Fl_Group(10, 10, 20, 20);
    window.add(v2);
    // do something else...
    window.remove(v2);
    delete v2;
#if FLTK_CASE_88_WORKAROUND
    Fl_Group::current(NULL);
#endif
    // Otherwise, invalid memory could be referenced
    assert(Fl_Group::current() != v2);
    Fl_Group * v3 = new Fl_Group(70, 70, 20, 20);
    window.add(v3);
    window.show();
    return Fl::run();
} // main()
```

Sorry, we should have probably simplified it to that to begin with.

**Comment by Albrecht-S** [comment]
Sorry for the confusion. My remark about the **difference** of your example concerned the way you delete widgets WRT the event handling loop vs. the way we were doing it. In our case it was done directly in the callback which could result in deleting the widget that got the event in its own callback. This is a special case, and in the old FLTK version it could happen that FLTK accessed the widget after the callback calling `clear_changed()` on it (and maybe more).

**Comment by Albrecht-S** [comment]
Thanks for the new, simplified example. It's obvious that the current code can leave the `Fl_Group::current_` variable pointing at a deleted widget and this needs to be fixed.

Side notes (nitpicking mode on): I hope the following remarks are helpful, but they are _not related_ to the issue at hand. I know it's example code, but there are some noteworthy aspects in case someone reads this later...

`window.add(v2);` is superfluous (but harmless) in this example since (as you know) FLTK calls `begin()` in the `Fl_Window` constructor. It's necessary though if you add `this->end();` or `Fl_Group::current(0);` in your derived (window) class.

`window.remove(v2); delete v2;` : again, the first statement is useless. The destructor would do this for you. This is documented behavior.

`Fl_Group * v3 = new Fl_Group(...); window.add(v3);` : The first statement would add the new group (`v3`) to any group if `Fl_Group::current_` is not `null` (and points at a valid `Fl_Group` widget). `window.add()` would remove it again from its current `parent()` group (if any) and add it to the desired group (window). In your real world program I would either make sure to call `Fl_Group::current(window);` before creating the new widget (group) and not add it explicitly or **always** call `Fl_Group::current(0);` _explicitly_ before creating a widget that is intended to be `add`ed to an arbitrary group in your program. This would avoid the in-and-out scenario described above in the first place. I believe you're doing the latter now, but anyway: I wrote this for the sake of completeness and for others reading this.

**Comment by Albrecht-S** [comment]
See [this file](https://github.com/fltk/fltk/files/4776850/issue-88-test3.cxx.txt) for a modified demo program including a patch proposal (rename to issue-88-test3.cxx).
The patch is included and you can feed the entire file to `patch -p1` (in the FLTK root dir) for simplicity.
I removed the assert() statement so you can "see it in action" with or w/o the fault. I used `valgrind` (under Linux) to show the issue instead.

Note: the patch includes two options, as proposed by @turix2: either execute `Fl_Group::current(0);` or `Fl_Group::current(parent());`.

Meanwhile I tend to use the latter because it's more consistent with FLTK's behavior in other cases. But I'd like to discuss this in [fltk.coredev](https://groups.google.com/forum/#!forum/fltkcoredev) before I commit it. Please follow the discussion [in thread: "GitHub Issue 88"](https://groups.google.com/d/msg/fltkcoredev/2GiPtTDV2GQ/AFtaQUnqBwAJ).

**Comment by Albrecht-S** [comment]
Fixed in a7a1e1594c7c574b135a99f80edbf47a7352e974, using solution (2) which makes the parent() of the deleted widget the current group. The FLTK team voted with 3:0 for this solution.

Thanks again for the report and for all contributions.
